rust/src/ldap/detect.rs

```diff
@@ -263,6 +274,139 @@ unsafe extern "C" fn ldap_detect_responses_count_free(_de: *mut c_void, ctx: *mu
+                let i32_index = i32::from_str(parts[1]).ok()?;
+                LdapIndex::Index(i32_index)
+            }
+        }
+    } else {
+        LdapIndex::Any
+    };
+    if let Some(ctx) = detect_parse_uint_enum::<u32, LdapResultCode>(parts[0]) {
+        let du32 = ctx;
+        return Some(DetectLdapRespResultData { du32, index });
+    }
+    return None;
+}
+
+unsafe extern "C" fn ldap_parse_responses_result_code(
+    ustr: *const std::os::raw::c_char,
+) -> *mut DetectUintData<u32> {
+    let ft_name: &CStr = CStr::from_ptr(ustr); //unsafe
+    if let Ok(s) = ft_name.to_str() {
+        if let Some(ctx) = aux_ldap_parse_resp_result_code(s) {
+            let boxed = Box::new(ctx);
+            return Box::into_raw(boxed) as *mut _;
+        }
+    }
+    return std::ptr::null_mut();
+}
+
+unsafe extern "C" fn ldap_detect_responses_result_code_setup(
+    de: *mut c_void, s: *mut c_void, raw: *const libc::c_char,
+) -> c_int {
+    if DetectSignatureSetAppProto(s, ALPROTO_LDAP) != 0 {
+        return -1;
+    }
+    let ctx = ldap_parse_responses_result_code(raw) as *mut c_void;
+    if ctx.is_null() {
+        return -1;
+    }
+    if SigMatchAppendSMToList(
+        de,
+        s,
+        G_LDAP_RESPONSES_RESULT_CODE_KW_ID,
+        ctx,
+        G_LDAP_RESPONSES_RESULT_CODE_BUFFER_ID,
+    )
+    .is_null()
+    {
+        ldap_detect_responses_result_code_free(std::ptr::null_mut(), ctx);
+        return -1;
+    }
+    return 0;
+}
+
+fn get_ldap_result_code(response: &LdapMessage) -> Option<u32> {
+    let result_code = match &response.protocol_op {
+        ProtocolOp::BindResponse(req) => Some(req.result.result_code.0),
+        ProtocolOp::SearchResultDone(req) => Some(req.result_code.0),
+        ProtocolOp::ModifyResponse(req) => Some(req.result.result_code.0),
+        ProtocolOp::AddResponse(req) => Some(req.result_code.0),
+        ProtocolOp::DelResponse(req) => Some(req.result_code.0),
+        ProtocolOp::ModDnResponse(req) => Some(req.result_code.0),
+        ProtocolOp::CompareResponse(req) => Some(req.result_code.0),
+        ProtocolOp::ExtendedResponse(req) => Some(req.result.result_code.0),
+        _ => None,
+    };
+    return result_code;
+}
+
+unsafe extern "C" fn ldap_detect_responses_result_code_match(
+    _de: *mut c_void, _f: *mut c_void, _flags: u8, _state: *mut c_void, tx: *mut c_void,
+    _sig: *const c_void, ctx: *const c_void,
+) -> c_int {
+    let tx = cast_pointer!(tx, LdapTransaction);
+    let ctx = cast_pointer!(ctx, DetectLdapRespResultData);
+
+    match ctx.index {
+        LdapIndex::Any => {
+            for responses in &tx.responses {
+                let result_code = get_ldap_result_code(responses);
+                if let Some(rc) = result_code {
+                    if rs_detect_u32_match(rc, &ctx.du32) == 1 {
+                        return 1;
+                    }
+                }
+            }
+            return 0;
+        }
+        LdapIndex::All => {
```

Comment:
I wonder if you can make this more generic so that you can say `return match_at_index(ctx.index, tx.responses as iterator with get_ldap_result_code, rs_detect_u32_match)`
Do you see what I mean ?